File: AIiOS/AIiOS/Services/GeminiService.swift
import Foundation

/// Calls Google's Gemini API — the cloud tier of the LLM router.
/// Takes the same `InferenceMessage` shape as `InferenceService` so call
/// sites (and eventually `RouterService`) can swap local/cloud without
/// reshaping the conversation.
final class GeminiService {

    static let shared = GeminiService()
    private init() {}

    /// "latest" alias so we don't need to track Google's dot-version bumps.
    static let defaultModel = "gemini-flash-latest"

    private let session = URLSession.shared

    enum GeminiServiceError: LocalizedError {
        case missingApiKey
        case httpError(Int, String)
        case emptyResponse
        case decodingError(String)

        var errorDescription: String? {
            switch self {
            case .missingApiKey:
                return "No Gemini API key set. Add one in Settings → Cloud AI."
            case .httpError(let code, let body):
                return "Gemini API error \(code): \(body)"
            case .emptyResponse:
                return "Gemini returned an empty response."
            case .decodingError(let msg):
                return "Failed to decode Gemini response: \(msg)"
            }
        }
    }

    /// Non-streaming chat completion against Gemini.
    func chat(
        messages:     [InferenceMessage],
        systemPrompt: String? = nil,
        model:        String = GeminiService.defaultModel,
        maxTokens:    Int    = 2048,
        temperature:  Float  = 0.8
    ) async throws -> String {
        let apiKey = SettingsService.shared.geminiApiKey
        guard !apiKey.isEmpty else { throw GeminiServiceError.missingApiKey }

        let url = URL(string: "https://generativelanguage.googleapis.com/v1beta/models/\(model):generateContent")!

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 60
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        // Header (not query string) keeps the key out of URLs/logs.
        request.setValue(apiKey, forHTTPHeaderField: "x-goog-api-key")

        let body = GeminiRequest(
            contents: messages.map(GeminiContent.init),
            systemInstruction: systemPrompt.map { GeminiSystemInstruction(parts: [GeminiPart(text: $0)]) },
            generationConfig: GeminiGenerationConfig(temperature: temperature, maxOutputTokens: maxTokens)
        )
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await session.data(for: request)

        guard let http = response as? HTTPURLResponse else {
            throw GeminiServiceError.httpError(0, "no HTTP response")
        }
        guard http.statusCode == 200 else {
            let bodyStr = String(data: data, encoding: .utf8) ?? ""
            throw GeminiServiceError.httpError(http.statusCode, bodyStr)
        }

        let decoded: GeminiResponse
        do {
            decoded = try JSONDecoder().decode(GeminiResponse.self, from: data)
        } catch {
            throw GeminiServiceError.decodingError(error.localizedDescription)
        }

        guard let text = decoded.candidates?.first?.content?.parts?.first?.text else {
            throw GeminiServiceError.emptyResponse
        }
        return text
    }
}

// MARK: - Request models

private struct GeminiRequest: Encodable {
    let contents:           [GeminiContent]
    let systemInstruction:  GeminiSystemInstruction?
    let generationConfig:   GeminiGenerationConfig
}

private struct GeminiSystemInstruction: Encodable {
    let parts: [GeminiPart]
}

private struct GeminiGenerationConfig: Encodable {
    let temperature:     Float
    let maxOutputTokens: Int
}

private struct GeminiContent: Encodable {
    let role:  String
    let parts: [GeminiPart]

    init(_ message: InferenceMessage) {
        // Gemini uses "model" for assistant turns, "user" for everything else.
        role = message.role == "assistant" ? "model" : "user"

        var parts: [GeminiPart] = []
        if !message.text.isEmpty {
            parts.append(GeminiPart(text: message.text))
        }
        if let imageData = message.imageData {
            parts.append(GeminiPart(inlineData: GeminiInlineData(
                mimeType: "image/jpeg",
                data: imageData.base64EncodedString()
            )))
        }
        self.parts = parts
    }
}

/// Exactly one of `text` / `inlineData` is set per part — custom encode(to:)
/// keeps the unused one out of the JSON body entirely (rather than `null`).
private struct GeminiPart: Encodable {
    var text:       String?
    var inlineData: GeminiInlineData?

    init(text: String)               { self.text = text }
    init(inlineData: GeminiInlineData) { self.inlineData = inlineData }

    enum CodingKeys: String, CodingKey { case text, inlineData }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encodeIfPresent(text, forKey: .text)
        try c.encodeIfPresent(inlineData, forKey: .inlineData)
    }
}

private struct GeminiInlineData: Encodable {
    let mimeType: String
    let data:     String
}

// MARK: - Response models

private struct GeminiResponse: Decodable {
    let candidates: [GeminiCandidate]?
}

private struct GeminiCandidate: Decodable {
    let content:      GeminiResponseContent?
    let finishReason: String?
}

private struct GeminiResponseContent: Decodable {
    let parts: [GeminiResponsePart]?
    let role:  String?
}

private struct GeminiResponsePart: Decodable {
    let text: String?
}

File: AIAndroid/app/src/main/kotlin/ai/ecoinference/app/services/GeminiService.kt
package ai.ecoinference.app.services

import ai.ecoinference.app.inference.InferenceMessage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.Base64
import java.util.concurrent.TimeUnit

class GeminiServiceException(message: String) : Exception(message)

/**
 * Calls Google's Gemini API — the cloud tier of the LLM router.
 * Takes the same [InferenceMessage] shape InferenceService uses so call sites
 * can swap local/cloud without reshaping the conversation. Mirrors iOS
 * GeminiService.swift.
 */
object GeminiService {

    /** "latest" alias so we don't need to track Google's dot-version bumps. */
    const val DEFAULT_MODEL = "gemini-flash-latest"

    private val client = OkHttpClient.Builder()
        .connectTimeout(60, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .build()

    private val json = Json { ignoreUnknownKeys = true; explicitNulls = false }

    /** Non-streaming chat completion against Gemini. */
    suspend fun chat(
        messages: List<InferenceMessage>,
        systemPrompt: String? = null,
        apiKey: String,
        model: String = DEFAULT_MODEL,
        maxTokens: Int = 2048,
        temperature: Float = 0.8f,
    ): String = withContext(Dispatchers.IO) {
        if (apiKey.isBlank()) {
            throw GeminiServiceException("No Gemini API key set. Add one in Settings → Cloud AI.")
        }

        val body = GeminiRequest(
            contents = messages.map { it.toGeminiContent() },
            systemInstruction = systemPrompt?.takeIf { it.isNotBlank() }
                ?.let { GeminiSystemInstruction(listOf(GeminiPart(text = it))) },
            generationConfig = GeminiGenerationConfig(temperature, maxTokens),
        )
        val requestJson = json.encodeToString(body)

        val request = Request.Builder()
            .url("https://generativelanguage.googleapis.com/v1beta/models/$model:generateContent")
            .addHeader("Content-Type", "application/json")
            // Header (not query string) keeps the key out of URLs/logs.
            .addHeader("x-goog-api-key", apiKey)
            .post(requestJson.toRequestBody("application/json".toMediaType()))
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw GeminiServiceException("Gemini API error ${response.code}: $responseBody")
            }
            val parsed: GeminiResponse = try {
                json.decodeFromString(responseBody)
            } catch (e: Exception) {
                throw GeminiServiceException("Failed to decode Gemini response: ${e.message}")
            }
            parsed.candidates?.firstOrNull()?.content?.parts?.firstOrNull()?.text
                ?: throw GeminiServiceException("Gemini returned an empty response.")
        }
    }
}

private fun InferenceMessage.toGeminiContent(): GeminiContent {
    // Gemini uses "model" for assistant turns, "user" for everything else.
    val role = if (this.role == "assistant") "model" else "user"
    val parts = buildList {
        if (text.isNotEmpty()) add(GeminiPart(text = text))
        imageBytes?.let {
            add(GeminiPart(inlineData = GeminiInlineData(
                mimeType = "image/jpeg",
                data     = Base64.getEncoder().encodeToString(it),
            )))
        }
    }
    return GeminiContent(role, parts)
}

// MARK: - Request models

@Serializable
private data class GeminiRequest(
    val contents: List<GeminiContent>,
    val systemInstruction: GeminiSystemInstruction? = null,
    val generationConfig: GeminiGenerationConfig,
)

@Serializable
private data class GeminiSystemInstruction(val parts: List<GeminiPart>)

@Serializable
private data class GeminiGenerationConfig(val temperature: Float, val maxOutputTokens: Int)

@Serializable
private data class GeminiContent(val role: String, val parts: List<GeminiPart>)

/** Exactly one of [text] / [inlineData] is set per part. */
@Serializable
private data class GeminiPart(
    val text: String? = null,
    val inlineData: GeminiInlineData? = null,
)

@Serializable
private data class GeminiInlineData(val mimeType: String, val data: String)

// MARK: - Response models

@Serializable
private data class GeminiResponse(val candidates: List<GeminiCandidate>? = null)

@Serializable
private data class GeminiCandidate(val content: GeminiResponseContent? = null)

@Serializable
private data class GeminiResponseContent(val parts: List<GeminiResponsePart>? = null)

@Serializable
private data class GeminiResponsePart(val text: String? = null)
